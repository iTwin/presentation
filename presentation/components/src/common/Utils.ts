/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Core */

import * as React from "react";
import { I18NNamespace } from "@bentley/imodeljs-i18n";
import { Presentation } from "@bentley/presentation-frontend";

/**
 * An interface of something that has a priority.
 * @internal
 */
export interface IPrioritized {
  priority: number;
}

/**
 * A sorting algorithm for `Array.sort` that sorts items by priority.
 * Higher priority items appear first in the list.
 *
 * @internal
 */
export const prioritySortFunction = (a: IPrioritized, b: IPrioritized): number => {
  if (a.priority > b.priority)
    return -1;
  if (a.priority < b.priority)
    return 1;
  return 0;
};

let localizationNamespace: I18NNamespace | undefined;
/**
 * Translate a string with the specified id from `PresentationComponents`
 * localization namespace. The `stringId` should not contain namespace - it's
 * prepended automatically.
 *
 * @internal
 */
export const translate = async (stringId: string): Promise<string> => {
  const localizationNamespaceName = "PresentationComponents";
  if (!localizationNamespace) {
    localizationNamespace = Presentation.i18n.registerNamespace(localizationNamespaceName);
  }
  await localizationNamespace.readFinished;
  stringId = `${localizationNamespaceName}:${stringId}`;
  return Presentation.i18n.translate(stringId);
};

/**
 * Creates a display name for the supplied component
 * @internal
 */
export const getDisplayName = <P>(component: React.ComponentType<P>): string => {
  if (component.displayName)
    return component.displayName;
  if (component.name)
    return component.name;
  return "Component";
};
