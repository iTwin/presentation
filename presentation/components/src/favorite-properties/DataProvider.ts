/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @module FavoriteProperties */

import * as _ from "lodash";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { CategoryDescription } from "@bentley/presentation-common";
import { IElementPropertyDataProvider, PropertyData } from "@bentley/ui-components";
import { Presentation, getScopeId } from "@bentley/presentation-frontend";
import { translate } from "../common/Utils";
import { PresentationPropertyDataProvider } from "../propertygrid/DataProvider";

const favoritesCategoryName = "Favorite";
let favoritesCategoryPromise: Promise<CategoryDescription> | undefined;

/** @internal */
export const getFavoritesCategory = async (): Promise<CategoryDescription> => {
  if (!favoritesCategoryPromise) {
    favoritesCategoryPromise = Promise.all([
      translate("categories.favorite.label"),
      translate("categories.favorite.description"),
    ]).then(([label, description]): CategoryDescription => ({
      name: favoritesCategoryName,
      label,
      description,
      priority: Number.MAX_VALUE,
      expand: true,
    }));
  }
  return favoritesCategoryPromise;
};

/** @beta */
export interface FavoritePropertiesDataProviderProps {
  /** @internal */
  propertyDataProviderFactory?: (imodel: IModelConnection, rulesetId?: string) => PresentationPropertyDataProvider;
}

/**
 * Presentation Rules-driven element favorite properties data provider implementation.
 * @beta
 */
export class FavoritePropertiesDataProvider implements IElementPropertyDataProvider {

  /**
   * Should fields with no values be included in the property list. No value means:
   * - For *primitive* fields: null, undefined, "" (empty string)
   * - For *array* fields: [] (empty array)
   * - For *struct* fields: {} (object with no members)
   */
  public includeFieldsWithNoValues: boolean;

  /**
   * Should fields with composite values be included in the property list.
   * Fields with composite values:
   * - *array* fields.
   * - *struct* fields.
   */
  public includeFieldsWithCompositeValues: boolean;

  /**
   * ID of a custom ruleset to use when requesting properties.
   */
  public customRulesetId: string | undefined;

  private _propertyDataProviderFactory: (imodel: IModelConnection, rulesetId?: string) => PresentationPropertyDataProvider;

  /** Constructor. */
  constructor(props?: FavoritePropertiesDataProviderProps) {
    this.includeFieldsWithNoValues = true;
    this.includeFieldsWithCompositeValues = true;
    this._propertyDataProviderFactory = props && props.propertyDataProviderFactory ?
      props.propertyDataProviderFactory : /* istanbul ignore next */
      (imodel: IModelConnection, rulesetId?: string) => new PresentationPropertyDataProvider(imodel, rulesetId);
  }

  /**
   * Returns PropertyData for the specified element.
   * PropertyData only contains a single category for favorite properties (if there are any).
   */
  public async getData(imodel: IModelConnection, elementId: string): Promise<PropertyData> {
    const key = await Presentation.selection.scopes.computeSelection(
      imodel,
      elementId,
      getScopeId(Presentation.selection.scopes.activeScope));

    const propertyDataProvider = this._propertyDataProviderFactory(imodel, this.customRulesetId);
    propertyDataProvider.keys = key;
    propertyDataProvider.includeFieldsWithNoValues = this.includeFieldsWithNoValues;
    propertyDataProvider.includeFieldsWithCompositeValues = this.includeFieldsWithCompositeValues;
    const propertyData = await propertyDataProvider.getData();

    // leave only favorite properties
    const favoritesCategory = await getFavoritesCategory();
    propertyData.categories = propertyData.categories.filter((c) => c.name === favoritesCategory.name);
    propertyData.records = propertyData.records.hasOwnProperty(favoritesCategory.name) ?
      { [favoritesCategory.name]: propertyData.records[favoritesCategory.name] } : {};

    return propertyData;
  }
}
