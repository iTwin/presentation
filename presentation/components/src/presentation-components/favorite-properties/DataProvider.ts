/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module FavoriteProperties
 */

import * as _ from "lodash";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { CategoryDescription } from "@bentley/presentation-common";
import { IElementPropertyDataProvider, PropertyData } from "@bentley/ui-components";
import { Presentation, getScopeId } from "@bentley/presentation-frontend";
import { translate } from "../common/Utils";
import { PresentationPropertyDataProvider } from "../propertygrid/DataProvider";

/** @internal */
export const FAVORITES_CATEGORY_NAME = "Favorite";

/** @internal */
export const getFavoritesCategory = (): CategoryDescription => {
  return {
    name: FAVORITES_CATEGORY_NAME,
    label: translate("categories.favorite.label"),
    description: translate("categories.favorite.description"),
    priority: Number.MAX_VALUE,
    expand: true,
  };
};

/** @beta */
export interface FavoritePropertiesDataProviderProps {
  /** @internal */
  propertyDataProviderFactory?: (imodel: IModelConnection, rulesetId?: string) => PresentationPropertyDataProvider;
}

/**
 * Presentation Rules-driven element favorite properties data provider implementation.
 * @beta
 */
export class FavoritePropertiesDataProvider implements IElementPropertyDataProvider {

  /**
   * Should fields with no values be included in the property list. No value means:
   * - For *primitive* fields: null, undefined, "" (empty string)
   * - For *array* fields: [] (empty array)
   * - For *struct* fields: {} (object with no members)
   */
  public includeFieldsWithNoValues: boolean;

  /**
   * Should fields with composite values be included in the property list.
   * Fields with composite values:
   * - *array* fields.
   * - *struct* fields.
   */
  public includeFieldsWithCompositeValues: boolean;

  /**
   * ID of a custom ruleset to use when requesting properties.
   */
  public customRulesetId: string | undefined;

  private _propertyDataProviderFactory: (imodel: IModelConnection, rulesetId?: string) => PresentationPropertyDataProvider;

  /** Constructor. */
  constructor(props?: FavoritePropertiesDataProviderProps) {
    this.includeFieldsWithNoValues = true;
    this.includeFieldsWithCompositeValues = true;
    this._propertyDataProviderFactory = props && props.propertyDataProviderFactory ?
      props.propertyDataProviderFactory : /* istanbul ignore next */
      (imodel: IModelConnection, rulesetId?: string) => new PresentationPropertyDataProvider(imodel, rulesetId);
  }

  /**
   * Returns PropertyData for the specified element.
   * PropertyData only contains a single category for favorite properties (if there are any).
   */
  public async getData(imodel: IModelConnection, elementId: string): Promise<PropertyData> {
    const key = await Presentation.selection.scopes.computeSelection(
      imodel,
      elementId,
      getScopeId(Presentation.selection.scopes.activeScope));

    const propertyDataProvider = this._propertyDataProviderFactory(imodel, this.customRulesetId);
    propertyDataProvider.keys = key;
    propertyDataProvider.includeFieldsWithNoValues = this.includeFieldsWithNoValues;
    propertyDataProvider.includeFieldsWithCompositeValues = this.includeFieldsWithCompositeValues;
    const propertyData = await propertyDataProvider.getData();

    // leave only favorite properties
    const favoritesCategory = getFavoritesCategory();
    propertyData.categories = propertyData.categories.filter((c) => c.name === favoritesCategory.name);
    propertyData.records = propertyData.records.hasOwnProperty(favoritesCategory.name) ?
      { [favoritesCategory.name]: propertyData.records[favoritesCategory.name] } : {};

    return propertyData;
  }
}
