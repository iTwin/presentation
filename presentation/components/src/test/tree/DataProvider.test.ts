/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/* tslint:disable:no-direct-imports */

import "@bentley/presentation-frontend/lib/test/_helpers/MockFrontendEnvironment";
import { expect } from "chai";
import * as faker from "faker";
import * as sinon from "sinon";
import { Logger } from "@bentley/bentleyjs-core";
import * as moq from "@bentley/presentation-common/lib/test/_helpers/Mocks";
import { PromiseContainer } from "@bentley/presentation-common/lib/test/_helpers/Promises";
import { NodesResponse, Node } from "@bentley/presentation-common";
import { createRandomECInstanceNodeKey, createRandomECInstanceNode, createRandomNodePathElement } from "@bentley/presentation-common/lib/test/_helpers/random";
import { createRandomTreeNodeItem } from "../_helpers/UiComponents";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { Presentation, PresentationManager } from "@bentley/presentation-frontend";
import { PageOptions } from "@bentley/ui-components";
import { PresentationTreeDataProvider } from "../../tree/DataProvider";
import { pageOptionsUiToPresentation } from "../../tree/Utils";

describe("TreeDataProvider", () => {

  let rulesetId: string;
  let provider: PresentationTreeDataProvider;
  const presentationManagerMock = moq.Mock.ofType<PresentationManager>();
  const imodelMock = moq.Mock.ofType<IModelConnection>();
  before(() => {
    rulesetId = faker.random.word();
    Presentation.presentation = presentationManagerMock.object;
  });
  beforeEach(() => {
    presentationManagerMock.reset();
    provider = new PresentationTreeDataProvider(imodelMock.object, rulesetId);
  });

  describe("rulesetId", () => {

    it("returns rulesetId provider is initialized with", () => {
      expect(provider.rulesetId).to.eq(rulesetId);
    });

  });

  describe("imodel", () => {

    it("returns imodel provider is initialized with", () => {
      expect(provider.imodel).to.eq(imodelMock.object);
    });

  });

  describe("getNodesCount", () => {

    describe("root", () => {

      it("returns presentation manager result", async () => {
        const resultNodes = [createRandomECInstanceNode(), createRandomECInstanceNode()];
        presentationManagerMock
          .setup((x) => x.getNodes({ imodel: imodelMock.object, rulesetId, paging: undefined }, undefined))
          .returns(async () => resultNodes)
          .verifiable();
        const actualResult = await provider.getNodesCount();
        expect(actualResult).to.eq(resultNodes.length);
        presentationManagerMock.verifyAll();
      });

      it("memoizes result", async () => {
        const resultNodes = [createRandomECInstanceNode(), createRandomECInstanceNode()];
        const resultContainers = [new PromiseContainer<NodesResponse>(), new PromiseContainer<NodesResponse>()];
        presentationManagerMock
          .setup((x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetId, paging: { start: 0, size: 10 } }, undefined))
          .returns(() => resultContainers[0].promise);
        presentationManagerMock
          .setup((x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetId, paging: { start: 0, size: 10 } }, undefined))
          .returns(() => resultContainers[1].promise);
        provider.pagingSize = 10;
        const promises = [provider.getNodesCount(), provider.getNodesCount()];
        resultContainers.forEach((c: PromiseContainer<NodesResponse>, index: number) => c.resolve({ nodes: resultNodes, count: index }));
        const results = await Promise.all(promises);
        expect(results[1]).to.eq(results[0]).to.eq(0);
        presentationManagerMock.verify((x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetId, paging: { start: 0, size: 10 } }, undefined), moq.Times.once());
      });

      it("requests count and first page when paging size is set", async () => {
        const result = { nodes: [createRandomECInstanceNode(), createRandomECInstanceNode()], count: 2 + faker.random.number() };
        const pageSize = 20;
        presentationManagerMock
          .setup((x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetId, paging: pageOptionsUiToPresentation({ start: 0, size: pageSize }) }, undefined))
          .returns(async () => result)
          .verifiable();
        provider.pagingSize = pageSize;
        const actualResult = await provider.getNodesCount();
        expect(actualResult).to.eq(result.count);
        presentationManagerMock.verifyAll();
      });

      it("returns nodes count equal to requested nodes list count when page options are undefined", async () => {
        const nodes = [createRandomECInstanceNode(), createRandomECInstanceNode(), createRandomECInstanceNode(), createRandomECInstanceNode()];
        presentationManagerMock.setup((x) => x.getNodes({ imodel: imodelMock.object, rulesetId, paging: undefined }, undefined))
          .returns(async () => nodes)
          .verifiable(moq.Times.once());
        presentationManagerMock.setup((x) => x.getNodesCount(moq.It.isAny(), moq.It.isAny()))
          .verifiable(moq.Times.never());
        const count = await provider.getNodesCount();
        presentationManagerMock.verifyAll();
        expect(count).to.equal(nodes.length);
      });
    });

    describe("children", () => {

      it("returns presentation manager result", async () => {
        const resultNodes = [createRandomECInstanceNode(), createRandomECInstanceNode()];
        const parentKey = createRandomECInstanceNodeKey();
        const parentNode = createRandomTreeNodeItem(parentKey);
        presentationManagerMock
          .setup((x) => x.getNodes({ imodel: imodelMock.object, rulesetId, paging: undefined }, parentKey))
          .returns(async () => resultNodes)
          .verifiable();
        const actualResult = await provider.getNodesCount(parentNode);
        expect(actualResult).to.eq(resultNodes.length);
        presentationManagerMock.verifyAll();
      });

      it("memoizes result", async () => {
        const parentKeys = [createRandomECInstanceNodeKey(), createRandomECInstanceNodeKey()];
        const parentNodes = parentKeys.map((key) => createRandomTreeNodeItem(key));
        const resultContainers = [new PromiseContainer<NodesResponse>(), new PromiseContainer<NodesResponse>()];

        presentationManagerMock
          .setup((x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetId, paging: { start: 0, size: 10 } }, parentKeys[0]))
          .returns(() => resultContainers[0].promise)
          .verifiable(moq.Times.once());
        presentationManagerMock
          .setup((x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetId, paging: { start: 0, size: 10 } }, parentKeys[1]))
          .returns(() => resultContainers[1].promise)
          .verifiable(moq.Times.once());

        provider.pagingSize = 10;
        const promises = [
          provider.getNodesCount(parentNodes[0]),
          provider.getNodesCount(parentNodes[1]),
          provider.getNodesCount(parentNodes[0]),
        ];
        resultContainers.forEach((c: PromiseContainer<NodesResponse>, index: number) => c.resolve({ nodes: [createRandomECInstanceNode(), createRandomECInstanceNode()], count: index }));
        const results = await Promise.all(promises);
        expect(results[0]).to.eq(results[2]).to.eq(0);
        expect(results[1]).to.eq(1);

        presentationManagerMock.verifyAll();
      });

      it("uses default page options", async () => {
        const parentKey = createRandomECInstanceNodeKey();
        const parentNode = createRandomTreeNodeItem(parentKey);
        const result = { nodes: [], count: faker.random.number() };
        presentationManagerMock
          .setup((x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetId, paging: pageOptionsUiToPresentation({ start: 0, size: 20 }) }, parentKey))
          .returns(async () => result)
          .verifiable();
        provider.pagingSize = 20;
        const actualResult = await provider.getNodesCount(parentNode);
        expect(actualResult).to.eq(result.count);
        presentationManagerMock.verifyAll();
      });

      it("returns nodes count equal to requested nodes list count when page options are undefined", async () => {
        const parentKey = createRandomECInstanceNodeKey();
        const parentNode = createRandomTreeNodeItem(parentKey);
        const nodes = [createRandomECInstanceNode(), createRandomECInstanceNode(), createRandomECInstanceNode(), createRandomECInstanceNode()];
        presentationManagerMock.setup((x) => x.getNodes({ imodel: imodelMock.object, rulesetId, paging: undefined }, parentKey))
          .returns(async () => nodes)
          .verifiable(moq.Times.once());
        presentationManagerMock.setup((x) => x.getNodesCount(moq.It.isAny(), moq.It.isAny()))
          .verifiable(moq.Times.never());
        const count = await provider.getNodesCount(parentNode);
        presentationManagerMock.verifyAll();
        expect(count).to.equal(nodes.length);
      });

    });

  });

  describe("getNodes", () => {

    describe("root", () => {

      it("returns presentation manager result", async () => {
        const pageOptions: PageOptions = { start: 0, size: faker.random.number() };
        const result = { nodes: [createRandomECInstanceNode(), createRandomECInstanceNode()], count: 2 };
        presentationManagerMock
          .setup((x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetId, paging: pageOptionsUiToPresentation(pageOptions) }, undefined))
          .returns(async () => result)
          .verifiable();
        const actualResult = await provider.getNodes(undefined, pageOptions);
        expect(actualResult).to.matchSnapshot();
        presentationManagerMock.verifyAll();
      });

      it("memoizes result", async () => {
        const resultnodesFirstPageContainer0 = new PromiseContainer<ReadonlyArray<Readonly<Node>>>();
        const resultnodesFirstPageContainer1 = new PromiseContainer<NodesResponse>();
        const resultNodesNonFirstPageContainer = new PromiseContainer<ReadonlyArray<Readonly<Node>>>();
        presentationManagerMock
          .setup((x) => x.getNodes({ imodel: imodelMock.object, rulesetId, paging: undefined }, undefined))
          .returns(() => resultnodesFirstPageContainer0.promise)
          .verifiable(moq.Times.once());
        presentationManagerMock
          .setup((x) => x.getNodes({ imodel: imodelMock.object, rulesetId, paging: { start: 0, size: 0 } }, undefined))
          .verifiable(moq.Times.never());
        presentationManagerMock
          .setup((x) => x.getNodes({ imodel: imodelMock.object, rulesetId, paging: { start: 1, size: 0 } }, undefined))
          .returns(() => resultNodesNonFirstPageContainer.promise)
          .verifiable(moq.Times.once());
        presentationManagerMock
          .setup((x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetId, paging: { start: 0, size: 1 } }, undefined))
          .returns(() => resultnodesFirstPageContainer1.promise)
          .verifiable(moq.Times.once());

        const promises = [
          provider.getNodes(), provider.getNodes(),
          provider.getNodes(undefined, { start: 0, size: 0 }), provider.getNodes(undefined, { start: 0, size: 0 }),
          provider.getNodes(undefined, { start: 1, size: 0 }), provider.getNodes(undefined, { start: 1, size: 0 }),
          provider.getNodes(undefined, { start: 0, size: 1 }), provider.getNodes(undefined, { start: 0, size: 1 }),
        ];
        resultnodesFirstPageContainer0.resolve([createRandomECInstanceNode()]);
        resultnodesFirstPageContainer1.resolve({ nodes: [createRandomECInstanceNode()], count: 1 });
        resultNodesNonFirstPageContainer.resolve([createRandomECInstanceNode()]);
        const results = await Promise.all(promises);

        expect(results[0]).to.eq(results[1], "results[0] should eq results[1]");
        expect(results[2])
          .to.eq(results[3], "results[2] should eq results[3]")
          .to.eq(results[0], "both results[2] and results[3] should eq results[0]");
        expect(results[4]).to.eq(results[5], "results[4] should eq results[5]");
        expect(results[6]).to.eq(results[7], "results[6] should eq results[7]");

        presentationManagerMock.verifyAll();
      });
    });

    describe("children", () => {

      it("returns presentation manager result", async () => {
        const parentKey = createRandomECInstanceNodeKey();
        const parentNode = createRandomTreeNodeItem(parentKey);
        const pageOptions: PageOptions = { start: 0, size: faker.random.number() };
        presentationManagerMock
          .setup((x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetId, paging: pageOptionsUiToPresentation(pageOptions) }, parentKey))
          .returns(async () => ({ nodes: [createRandomECInstanceNode(), createRandomECInstanceNode()], count: 2 }))
          .verifiable();
        const actualResult = await provider.getNodes(parentNode, pageOptions);
        expect(actualResult).to.matchSnapshot();
        presentationManagerMock.verifyAll();
      });

      it("memoizes result", async () => {
        const parentKeys = [createRandomECInstanceNodeKey(), createRandomECInstanceNodeKey()];
        const parentNodes = parentKeys.map((key) => createRandomTreeNodeItem(key));
        const resultnodesFirstPageContainer0 = new PromiseContainer<ReadonlyArray<Readonly<Node>>>();
        const resultnodesFirstPageContainer1 = new PromiseContainer<NodesResponse>();
        const resultNodesNonFirstPageContainer = new PromiseContainer<ReadonlyArray<Readonly<Node>>>();

        presentationManagerMock
          .setup((x) => x.getNodes({ imodel: imodelMock.object, rulesetId, paging: undefined }, parentKeys[0]))
          .returns(() => resultnodesFirstPageContainer0.promise)
          .verifiable(moq.Times.once());
        presentationManagerMock
          .setup((x) => x.getNodes({ imodel: imodelMock.object, rulesetId, paging: { start: 0, size: 0 } }, parentKeys[0]))
          .verifiable(moq.Times.never());
        presentationManagerMock
          .setup((x) => x.getNodes({ imodel: imodelMock.object, rulesetId, paging: { start: 1, size: 0 } }, parentKeys[0]))
          .returns(() => resultNodesNonFirstPageContainer.promise)
          .verifiable(moq.Times.once());
        presentationManagerMock
          .setup((x) => x.getNodesAndCount({ imodel: imodelMock.object, rulesetId, paging: { start: 0, size: 1 } }, parentKeys[1]))
          .returns(() => resultnodesFirstPageContainer1.promise)
          .verifiable(moq.Times.once());

        const promises = [
          provider.getNodes(parentNodes[0], undefined), provider.getNodes(parentNodes[0], undefined),
          provider.getNodes(parentNodes[0], { start: 0, size: 0 }), provider.getNodes(parentNodes[0], { start: 0, size: 0 }),
          provider.getNodes(parentNodes[0], { start: 1, size: 0 }), provider.getNodes(parentNodes[0], { start: 1, size: 0 }),
          provider.getNodes(parentNodes[1], { start: 0, size: 1 }), provider.getNodes(parentNodes[1], { start: 0, size: 1 }),
        ];
        resultnodesFirstPageContainer0.resolve([createRandomECInstanceNode()]);
        resultnodesFirstPageContainer1.resolve({ nodes: [createRandomECInstanceNode()], count: 1 });
        resultNodesNonFirstPageContainer.resolve([createRandomECInstanceNode()]);
        const results = await Promise.all(promises);

        expect(results[0]).to.eq(results[1], "results[0] should eq results[1]");
        expect(results[2])
          .to.eq(results[3], "results[2] should eq results[3]")
          .to.eq(results[0], "both results[2] and results[3] should eq results[0]");
        expect(results[4]).to.eq(results[5], "results[4] should eq results[5]");
        expect(results[6]).to.eq(results[7], "results[6] should eq results[7]");

        presentationManagerMock.verifyAll();
      });

    });

    it("Logs warning when requesting nodes and pagingSize is not the same as passed pageOptions", async () => {
      const pageOptions: PageOptions = { start: 0, size: 10 };
      const loggerSpy = sinon.spy(Logger, "logWarning");
      const result = { nodes: [createRandomECInstanceNode(), createRandomECInstanceNode()], count: 2 };
      presentationManagerMock.setup((x) => x.getNodesAndCount(moq.It.isAny(), moq.It.isAny())).returns(async () => result);
      presentationManagerMock.setup((x) => x.getNodes(moq.It.isAny(), moq.It.isAny())).returns(async () => result.nodes);

      // Paging size is not set and no pageOptions are passed
      await provider.getNodes();
      expect(loggerSpy.calledOnce).to.be.true;
      loggerSpy.resetHistory();

      // Paging size is not set and pageOptions are passed
      await provider.getNodes(undefined, pageOptions);
      expect(loggerSpy.calledOnce).to.be.true;
      loggerSpy.resetHistory();

      // Paging size is set and no pageOptions are passed
      provider.pagingSize = 10;
      await provider.getNodes();
      expect(loggerSpy.notCalled).to.be.true;
      loggerSpy.resetHistory();

      // Paging size is set and pageOptions are passed but not equal to paging size
      provider.pagingSize = 20;
      await provider.getNodes(undefined, pageOptions);
      expect(loggerSpy.calledOnce).to.be.true;
      loggerSpy.resetHistory();

      // Paging size is set and pageOptions are passed and equal to paging size
      provider.pagingSize = 10;
      await provider.getNodes(undefined, pageOptions);
      expect(loggerSpy.notCalled).to.be.true;
    });

  });

  describe("getFilteredNodes", () => {

    it("returns presentation manager result", async () => {
      const filter = faker.random.word();
      presentationManagerMock
        .setup((x) => x.getFilteredNodePaths({ imodel: imodelMock.object, rulesetId }, filter))
        .returns(async () => [createRandomNodePathElement(), createRandomNodePathElement()])
        .verifiable();
      const actualResult = await provider.getFilteredNodePaths(filter);
      expect(actualResult).to.matchSnapshot();
      presentationManagerMock.verifyAll();
    });

  });

});
