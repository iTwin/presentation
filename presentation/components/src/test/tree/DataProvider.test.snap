// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeDataProvider getFilteredNodes returns presentation manager result 1`] = `
Array [
  Object {
    "children": Array [],
    "index": 644,
    "isMarked": true,
    "node": Object {
      "backColor": "rgb(
    215,
    125,
    116
  )",
      "description": "Rerum laborum ipsa.",
      "foreColor": "#6EE32D",
      "hasChildren": true,
      "imageId": "Soft",
      "isCheckboxEnabled": false,
      "isCheckboxVisible": true,
      "isChecked": true,
      "isEditable": true,
      "isExpanded": false,
      "isSelectionDisabled": false,
      "key": Object {
        "instanceKeys": Array [
          Object {
            "className": "SQL",
            "id": "0x42e20000015b9b",
          },
          Object {
            "className": "viral",
            "id": "0x1b78000000b915",
          },
        ],
        "pathFromRoot": Array [
          "05b9034c-bb1b-47e5-a34b-ee893267e90a",
          "8528b854-ec02-4dd8-bfe4-a098a1d96013",
        ],
        "type": "ECInstancesNode",
      },
      "label": Object {
        "displayValue": "Auto Loan Account",
        "rawValue": "Awesome",
        "typeName": "string",
      },
    },
  },
  Object {
    "children": Array [],
    "index": 668,
    "node": Object {
      "backColor": undefined,
      "description": "Autem ipsa vel voluptatem quia recusandae.",
      "foreColor": "#2F04EF",
      "hasChildren": false,
      "imageId": "Chips",
      "isCheckboxEnabled": true,
      "isCheckboxVisible": false,
      "isChecked": false,
      "isEditable": false,
      "isExpanded": false,
      "isSelectionDisabled": false,
      "key": Object {
        "instanceKeys": Array [
          Object {
            "className": "Applications",
            "id": "0x6b0b000000b7ad",
          },
          Object {
            "className": "EXE",
            "id": "0xc6b4000001712e",
          },
        ],
        "pathFromRoot": Array [
          "a3944426-64d1-4202-bc22-44b3f7075dc2",
          "845797cd-fdf3-4462-ba83-b9a1ecd911f2",
        ],
        "type": "ECInstancesNode",
      },
      "label": Object {
        "displayValue": "Cordoba Oro",
        "rawValue": "purple",
        "typeName": "string",
      },
    },
  },
]
`;

exports[`TreeDataProvider getNodes children returns presentation manager result 1`] = `
Array [
  Object {
    "__presentation-components/key": Object {
      "instanceKeys": Array [
        Object {
          "className": "olive",
          "id": "0x69150000016f27",
        },
        Object {
          "className": "cross-platform",
          "id": "0xc11d0000006f7b",
        },
      ],
      "pathFromRoot": Array [
        "e5e66399-295a-454e-a1ab-590b334ae253",
        "b5928504-ca89-4d67-9fe9-01f13624cefb",
      ],
      "type": "ECInstancesNode",
    },
    "checkBoxState": 1,
    "description": "Sunt excepturi est animi.",
    "hasChildren": true,
    "icon": "motivating",
    "id": "b5928504-ca89-4d67-9fe9-01f13624cefb/e5e66399-295a-454e-a1ab-590b334ae253",
    "isCheckboxVisible": true,
    "label": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "Multi-channelled",
        "value": "deposit",
        "valueFormat": 0,
      },
    },
    "parentId": "b4da3271-b2d6-497b-a799-93a4436cec5f",
    "style": Object {
      "colorOverrides": Object {
        "color": -717867777,
      },
    },
  },
  Object {
    "__presentation-components/key": Object {
      "instanceKeys": Array [
        Object {
          "className": "Practical Concrete Pizza",
          "id": "0xc14b00000120e7",
        },
        Object {
          "className": "platforms",
          "id": "0x8a760000016461",
        },
      ],
      "pathFromRoot": Array [
        "377376d7-efbb-4df5-a70a-4764c60c82ae",
        "baa5384c-065b-43f7-bfbe-9d55e79bfb1f",
      ],
      "type": "ECInstancesNode",
    },
    "description": "Sunt eveniet facilis officia.",
    "hasChildren": true,
    "icon": "Technician",
    "id": "baa5384c-065b-43f7-bfbe-9d55e79bfb1f/377376d7-efbb-4df5-a70a-4764c60c82ae",
    "label": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "Antigua and Barbuda",
        "value": "Shoes",
        "valueFormat": 0,
      },
    },
    "parentId": "b4da3271-b2d6-497b-a799-93a4436cec5f",
    "style": Object {
      "colorOverrides": Object {
        "backgroundColor": 1215880959,
        "color": -1153090305,
      },
    },
  },
]
`;

exports[`TreeDataProvider getNodes root returns presentation manager result 1`] = `
Array [
  Object {
    "__presentation-components/key": Object {
      "instanceKeys": Array [
        Object {
          "className": "data-warehouse",
          "id": "0xc8340000009d04",
        },
        Object {
          "className": "Washington",
          "id": "0xc7df000000977f",
        },
      ],
      "pathFromRoot": Array [
        "48ae5ef5-0f6e-4931-ba46-85cc14189a13",
        "8287e887-4294-40a6-b4fc-635a36a5784a",
      ],
      "type": "ECInstancesNode",
    },
    "autoExpand": true,
    "icon": "Unbranded",
    "id": "8287e887-4294-40a6-b4fc-635a36a5784a/48ae5ef5-0f6e-4931-ba46-85cc14189a13",
    "label": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "payment",
        "value": "open-source",
        "valueFormat": 0,
      },
    },
  },
  Object {
    "__presentation-components/key": Object {
      "instanceKeys": Array [
        Object {
          "className": "backing up",
          "id": "0xe5bf0000011f41",
        },
        Object {
          "className": "auxiliary",
          "id": "0x14bc1000001057a",
        },
      ],
      "pathFromRoot": Array [
        "3c15e225-8d91-4174-8f69-1c86639b762a",
        "6c7177b4-13aa-482a-9d6b-76624dbc4a02",
      ],
      "type": "ECInstancesNode",
    },
    "autoExpand": true,
    "hasChildren": true,
    "id": "6c7177b4-13aa-482a-9d6b-76624dbc4a02/3c15e225-8d91-4174-8f69-1c86639b762a",
    "isCheckboxDisabled": true,
    "isCheckboxVisible": true,
    "label": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "Manager",
        "value": "Outdoors",
        "valueFormat": 0,
      },
    },
    "style": Object {
      "colorOverrides": Object {
        "color": 1986721023,
      },
    },
  },
]
`;
